sdk/core/System.ClientModel/src/Pipeline/ClientPipeline.cs

```diff
@@ -0,0 +1,227 @@
+﻿// Copyright (c) Microsoft Corporation. All rights reserved.
+// Licensed under the MIT License.
+
+using System.ClientModel.Internal;
+using System.Collections;
+using System.Collections.Generic;
+using System.Threading.Tasks;
+
+namespace System.ClientModel.Primitives;
+
+public sealed partial class ClientPipeline
+{
+    private readonly int _perCallIndex;
+    private readonly int _perTryIndex;
+    private readonly int _beforeTransportIndex;
+
+    private readonly ReadOnlyMemory<PipelinePolicy> _policies;
+    private readonly PipelineTransport _transport;
+
+    private ClientPipeline(ReadOnlyMemory<PipelinePolicy> policies, int perCallIndex, int perTryIndex, int beforeTransportIndex)
+    {
+        if (perCallIndex > perTryIndex) throw new ArgumentOutOfRangeException(nameof(perCallIndex), "perCallIndex cannot be greater than perTryIndex.");
```

Comment:
perCallIndex and perTryIndex appear to come from local variables that the user wouldn't be aware of. Is this can actually be thrown, we should use a better message. If it is purely for validating that we don't break this invariant while making changes, consider making it a Debug.Assert.

Reply:
Yeah, great callout.  Will fix and thanks!